src/mlpack/methods/ann/layer/neural_turing_machine.hpp

```diff
@@ -0,0 +1,273 @@
+/**
+ * @file neural_turing_machine.hpp
+ * @author Sumedh Ghaisas
+ *
+ * Definition of the Neural Turing Machine class.
+ *
+ * For more information, read the following paper:
+ *
+ * @code
+ * @article{graves2014neural,
+ *    title   = {Neural turing machines},
+ *    author  = {Graves, Alex and Wayne, Greg and Danihelka, Ivo},
+ *    journal = {arXiv preprint arXiv:1410.5401},
+ *    year    = {2014}
+ * }
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_NTM_HPP
+#define MLPACK_METHODS_ANN_LAYER_NTM_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include <boost/ptr_container/ptr_vector.hpp>
+
+#include "../visitor/delta_visitor.hpp"
+#include "../visitor/output_parameter_visitor.hpp"
+#include "../activation_functions/softplus_function.hpp"
+
+#include "layer_types.hpp"
+#include "add_merge.hpp"
+#include "sequential.hpp"
+
+#include "memory_head.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * In Neural Turing Mahine, the expressiveness of a neural network,
+ * (or controller) is increased by attaching a memory to it. The input to this
+ * modified neural network consists of original input and the read memory. In
+ * turn the controller decides which memory locations to read and which memory
+ * locations to write. The controller output is taken as the final output.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class NeuralTuringMachine
+{
+ public:
+  /**
+   * Create the Neural Turing Machine layer object using the specified
+   * parameters.
+   *
+   * @param inSize The number of input units.
+   * @param outSize The number of output units.
+   * @param numMem Number of memory locations to use.
+   * @param memSize Size of each memory location.
+   * @param shiftSize Circular shift rotation size.
+   * @param controller The controller network to use.
+   */
+  NeuralTuringMachine(const size_t inSize,
+                      const size_t outSize,
+                      const size_t numMem,
+                      const size_t memSize,
+                      const size_t shiftSize,
+                      LayerTypes controller);
```

Comment:
I think a destructor also needs to be added to loop over `network` and use a `DeleteVisitor()`.

Reply:
Sure thing...